File: androidApp/src/main/kotlin/com/humblesolutions/aromex/ui/login/LoginViewModel.kt
package com.humblesolutions.aromex.ui.login

import android.app.Application
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.humblesolutions.aromex.data.AndroidPreferencesRepository
import com.humblesolutions.aromex.data.FirebaseAuthRepository
import com.humblesolutions.aromex.data.FirestoreUserRepository
import com.humblesolutions.aromex.data.HttpCompanyDirectoryRepository
import com.humblesolutions.aromex.model.AuthenticatedSession
import com.humblesolutions.aromex.model.LoginError
import com.humblesolutions.aromex.model.LoginException
import com.humblesolutions.aromex.model.ResolvedCompany
import com.humblesolutions.aromex.usecase.LoginResult
import com.humblesolutions.aromex.usecase.LoginUseCase
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch

data class LoginUiState(
    val email: String = "",
    val password: String = "",
    val isSubmitting: Boolean = false,
    val error: LoginError? = null,
    /** Non-null while the user is on the choose-company step. */
    val candidates: List<ResolvedCompany>? = null,
    /** Filled once login succeeds. The host UI navigates to Home. */
    val authenticated: AuthenticatedSession? = null,
)

class LoginViewModel(application: Application) : AndroidViewModel(application) {

    private val prefs = AndroidPreferencesRepository(application)
    private val directory = HttpCompanyDirectoryRepository()
    private val authRepo = FirebaseAuthRepository(application)
    private val userRepo = FirestoreUserRepository(application)
    private val loginUseCase = LoginUseCase(directory, authRepo, userRepo)

    private val _uiState = MutableStateFlow(LoginUiState())
    val uiState: StateFlow<LoginUiState> = _uiState.asStateFlow()

    fun onEmailChange(value: String) {
        _uiState.update { it.copy(email = value, error = null) }
    }

    fun onPasswordChange(value: String) {
        _uiState.update { it.copy(password = value, error = null) }
    }

    fun onDismissError() {
        _uiState.update { it.copy(error = null) }
    }

    fun onSubmit() {
        val state = _uiState.value
        if (state.isSubmitting) return
        val email = state.email.trim()
        val password = state.password
        if (email.isEmpty() || password.isEmpty()) return

        _uiState.update { it.copy(isSubmitting = true, error = null) }
        viewModelScope.launch {
            try {
                when (val result = loginUseCase.execute(email, password)) {
                    is LoginResult.Success -> {
                        rememberLastLogin(email, result.authenticated.session.companyId)
                        _uiState.update {
                            it.copy(isSubmitting = false, authenticated = result.authenticated, candidates = null)
                        }
                    }
                    is LoginResult.NeedsCompanyChoice -> {
                        _uiState.update {
                            it.copy(isSubmitting = false, candidates = result.candidates)
                        }
                    }
                }
            } catch (le: LoginException) {
                _uiState.update { it.copy(isSubmitting = false, error = le.error) }
            } catch (t: Throwable) {
                _uiState.update {
                    it.copy(isSubmitting = false, error = LoginError.Unexpected(t.message ?: "unknown"))
                }
            }
        }
    }

    fun onChooseCompany(company: ResolvedCompany) {
        val state = _uiState.value
        if (state.isSubmitting) return
        val email = state.email.trim()
        val password = state.password
        _uiState.update { it.copy(isSubmitting = true, error = null) }
        viewModelScope.launch {
            try {
                val authenticated = loginUseCase.finishLogin(company, email, password)
                rememberLastLogin(email, authenticated.session.companyId)
                _uiState.update {
                    it.copy(isSubmitting = false, authenticated = authenticated, candidates = null)
                }
            } catch (le: LoginException) {
                _uiState.update { it.copy(isSubmitting = false, error = le.error, candidates = null) }
            } catch (t: Throwable) {
                _uiState.update {
                    it.copy(isSubmitting = false, error = LoginError.Unexpected(t.message ?: "unknown"))
                }
            }
        }
    }

    fun onCancelChooseCompany() {
        _uiState.update { it.copy(candidates = null) }
    }

    private suspend fun rememberLastLogin(email: String, companyId: String) {
        prefs.setLastSignedInEmail(email)
        prefs.setLastCompanyId(companyId)
    }
}

File: iosApp/iosApp/viewmodel/LoginViewModel.swift
import Foundation
import SharedLogic

@MainActor
final class LoginViewModel: ObservableObject {
    @Published var email: String = ""
    @Published var password: String = ""
    @Published private(set) var isSubmitting: Bool = false
    @Published private(set) var error: LoginError? = nil
    @Published private(set) var candidates: [ResolvedCompany]? = nil
    @Published private(set) var authenticated: AuthenticatedSession? = nil

    private let prefs: PreferencesRepository
    private let directory: CompanyDirectoryRepository
    private let authRepo: AuthRepository
    private let userRepo: UserRepository
    private let loginUseCase: LoginUseCase

    init() {
        let prefs = UserDefaultsPreferencesRepository()
        let directory = HttpCompanyDirectoryRepository()
        let authRepo = FirebaseAuthRepositoryImpl()
        let userRepo = FirestoreUserRepository()
        self.prefs = prefs
        self.directory = directory
        self.authRepo = authRepo
        self.userRepo = userRepo
        self.loginUseCase = LoginUseCase(directory: directory, auth: authRepo, user: userRepo)
    }

    func onEmailChange(_ value: String) {
        email = value
        error = nil
    }

    func onPasswordChange(_ value: String) {
        password = value
        error = nil
    }

    func onDismissError() {
        error = nil
    }

    func onSubmit() {
        if isSubmitting { return }
        let trimmedEmail = email.trimmingCharacters(in: .whitespacesAndNewlines)
        let pass = password
        if trimmedEmail.isEmpty || pass.isEmpty { return }

        isSubmitting = true
        error = nil
        Task {
            do {
                let result = try await loginUseCase.execute(email: trimmedEmail, password: pass)
                if let success = result as? LoginResult.Success {
                    try? await rememberLastLogin(email: trimmedEmail, companyId: success.authenticated.session.companyId)
                    isSubmitting = false
                    authenticated = success.authenticated
                    candidates = nil
                } else if let needsChoice = result as? LoginResult.NeedsCompanyChoice {
                    isSubmitting = false
                    candidates = needsChoice.candidates
                }
            } catch {
                isSubmitting = false
                self.error = mapLoginError(from: error)
            }
        }
    }

    func onChooseCompany(_ company: ResolvedCompany) {
        if isSubmitting { return }
        let trimmedEmail = email.trimmingCharacters(in: .whitespacesAndNewlines)
        let pass = password
        isSubmitting = true
        error = nil
        Task {
            do {
                let session = try await loginUseCase.finishLogin(
                    company: company,
                    email: trimmedEmail,
                    password: pass
                )
                try? await rememberLastLogin(email: trimmedEmail, companyId: session.session.companyId)
                isSubmitting = false
                authenticated = session
                candidates = nil
            } catch {
                isSubmitting = false
                self.error = mapLoginError(from: error)
                candidates = nil
            }
        }
    }

    // Kotlin exceptions arrive on Swift as NSError. When the underlying Kotlin
    // exception is a LoginException, Kotlin/Native tags it in userInfo under
    // the "KotlinException" key — unwrap it to get our typed LoginError.
    private func mapLoginError(from error: Error) -> LoginError {
        let ns = error as NSError
        if let ex = ns.userInfo["KotlinException"] as? LoginException {
            return ex.error
        }
        return LoginError.Unexpected(message: error.localizedDescription)
    }

    func onCancelChooseCompany() {
        candidates = nil
    }

    // Clears local state so the Login screen is fresh (used after sign-out).
    func reset() {
        email = ""
        password = ""
        isSubmitting = false
        error = nil
        candidates = nil
        authenticated = nil
    }

    private func rememberLastLogin(email: String, companyId: String) async throws {
        try await prefs.setLastSignedInEmail(email: email)
        try await prefs.setLastCompanyId(companyId: companyId)
    }
}
